SIPS/sip-20.md

```diff
@@ -1,189 +1,237 @@
+  }
 };
+```
+the type for an `onWebSocketEvent` handler function's arguments is:
 
-/**
- * A binary message received from a WebSocket.
- *
- * @property type - The type of the message.
- * @property message - The message as an ArrayBuffer.
- */
-type WebSocketBinaryMessage = {
-  type: 'binary';
-  message: ArrayBuffer;
+```typescript
+export type WebSocketMessageEvent = {
+  type: "message";
+  id: string;
+  origin: string;
+  dataType: "text" | "binary";
+  data: string | Uint8Array;
 };
 
-/**
- * A message received from a WebSocket.
- *
- * @property type - The type of the message.
- * @property message - The message as a string or an ArrayBuffer, depending on
- * the type.
- */
-type WebSocketData = {
-  type: 'websocket';
-  data: WebSocketTextMessage | WebSocketBinaryMessage;
+export type WebSocketOpenEvent = {
+  type: "open";
+  id: string;
+  origin: string;
 };
 
-/**
- * The data received from an external source.
- *
- * @property type - The type of the external data source.
- * @property data - The data received from the external data source.
- */
-type ExternalData = WebSocketData;
+export type WebSocketCloseEvent = {
+  type: "close";
+  id: string;
+  origin: string;
+};
 
-type OnExternalDataHandlerArgs = ExternalData & {
-  source: string;
+export type WebSocketErrorEvent = {
+  type: "error";
+  id: string;
+  origin: string;
 };
 
-type OnExternalDataHandler = (args: OnExternalDataHandlerArgs) => Promise<void>;
+export type WebSocketEvent =
+  | WebSocketDataEvent
+  | WebSocketOpenEvent
+  | WebSocketCloseEvent
+  | WebSocketErrorEvent;
 ```
 
-#### Example
-
 ```typescript
-import { OnExternalDataHandler } from '@metamask/snaps-types';
-import { assert } from '@metamask/utils';
+type OnWebSocketEventArgs = {
+  event: WebSocketEvent;
+};
+```
+`type` - The type of the WebSocket event, which can be one of the following:
+  - `message` - Indicates that a message has been received.
+  - `open` - Indicates that a WebSocket connection has been opened.
+  - `close` - Indicates that a WebSocket connection has been closed.
+  - `error` - Indicates that an error has occurred with the WebSocket connection.
 
-export const onExternalData: OnExternalDataHandler = ({ type, data, source }) => {
-  assert(type === 'websocket'); // `data` is inferred as `WebSocketData`.
-  assert(data.type === 'text'); // `message` is inferred as `string`.
+`id` - The unique identifier of the WebSocket connection associated with the event.
 
-  // Now the Snap can do whatever is needed with the message.
-  const json = JSON.parse(data.message);
-};
+`origin` - The origin of the WebSocket event.
+
+`dataType` - The type of data received in the event, which can be either `text` or `binary`. This property is only present for `message` events.
+
+`data` - The data received in the event. For `message` events, this can be either a string (for text messages) or a `Uint8Array` (for binary messages). For other event types, this property is not present.
```

Comment:
These shouldn't be documented here if they aren't present. Seems more clear to describe the different events separately.